<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>仿写repromise-master源码</title>
</head>
<body>

    <script type="text/javascript">

        /** # (1) Promise 代码基本结构 : 实例化 Promise 对象时传入一个函数 (resolver) 作为执行器, 执行器
         * 有2个函数参数 (resolve 和 reject) 分别将结果变为 "成功态 (fulfilled) " 或 "失败态 (rejected) "。 */
        class Promise {
            /*
             * executor /ɪg'zekjʊtə/ n. 执行者
             * execute  /'eksɪkjuːt/ vt. 执行
             * resolver /ri:'zɔlvə/  n. 分解器
             * Asynchrony (Async)  /ei'siŋkrəni/  n. 异步
             * synchrony (Sync) /'sɪŋkrənɪ/   n. 同步
             */
            constructor(resolver) {
                // 首先判断: resolver 是否存在并且是不是一个函数
                if (resolver && typeof resolver !== "function") {
                    throw new Error(`Promise resolver ${resolver} is not a function (参数 resolver 不是一个函数) `)
                }
                // 初始化 state 为等待态
                this.state = "pending";
                this.data = undefined;
                // 保存回调函数的数组
                this.callbackArr = [];
                // 如果执行者 resolver 是函数就立即调用当前构造函数的 executeResolve() 方法执行 resolver
                if (typeof resolver === "function") {
                    this.executeResolve(resolver);
                }
            }

            executeResolve(resolver) {
                const that = this;
                // 定义一个保存回调函数的变量 cb 默认值为 false
                let cb = false;
                // 定义执行成功时的 onSuccess 函数: 也即是 (符合规范的Promise-0.html) 中的 resolve() 函数,
                const onSuccess = function(value) {
                    // console.log("value",value);
                    // console.log("cb", cb);

                    // 这个 if 的意思大概时 cb 不存在的时候直接 return
                    if (cb) return;
                    cb = true;
                    // execute callback 执行回调函数，传入2个参数
                    that.executeCallback("fulfilled", value);
                };
                // 定义当执行失败时用来报错的 reject() 函数
                const onError = function(value) {
                    if (cb) return;
                    cb = true;
                    that.executeCallback("rejected", value)
                };

                // try{}catch(){} 会在当前方法发调用时立即执行
                try {
                    // 把上面定义的 onSuccess 函数和 onError 函数的"指针"传入到 resolver 执行器函数内，
                    // 注: 传入函数的指针，不会调用函数
                    resolver(onSuccess, onError);
                } catch (e) {
                    onError(e);
                }
            }


            /** 注: executeCallback 是执行同步回调函数，即 new Promise((resolve, reject) =>
             * { resolve(0) }) 实例内不含有 setTimeout() 超时调用的情况  */
            // 执行 new Promise() 实例中的 resolve(0), reject(1);
            // 注:
            // 参数: (1) type = "fulfilled / rejected", value
            //      (2) value 为 resolve() 或 reject() 调用的参数 (注: value 复杂点也有可能是 Promise.resolve())
            executeCallback(type, value) {
                // console.log("executeCallBack value parameter", value);
                const isResolve = type === "fulfilled";
                let thenable;
                if (isResolve && (Object.prototype.toString.call(value) === "[object Object]") ||
                    (typeof (value) === "function")) {
                    try {
                        thenable = this.getThen(value);
                    } catch (e) {
                        return  this.executeCallback.call(this, "rejectd", e)
                    }
                }
                // 如果是 thenable 对象而且是 fulfilled 状态 (Promise.reject() 会返回参数值)
                if (isResolve && thenable) {
                    // 最终会将 thenable 对象里的值*个抽出到 this.data 中
                    this.executeResolve(thenable);
                }
                // promise 状态一旦改变便不可更改
                else if (this.state === "pending") {
                    this.state = isResolve ? "fulfilled" : "rejected";
                    this.data = value;
                    this.callbackArr.forEach(fn => fn[type](value)) // ④
                }
                // 直接调用 Promise.resolve() Promise.reject() 用得到
                return this;
            }


            /** 注: execute Async callback 执行异步调用， 即这种含有超时调用的情况: new Promise(
             * (resolve, reject) => { setTimeout(() => { resolve(0) }, 500) } ) */
            executeAsyncCallback(callback, value) {
                const that = this;
                setTimeout(() => {
                   let res;
                   try {
                       res = callback(value);
                   } catch(e) {
                       // 目的: 使捕获到的错误进入 Promise.catch() 中
                       that.executeCallback("rejected", e);
                   }
                   // 2.3.1 如果放回的值与原 promise 相等，则使无穷循环
                    if (res === that) {
                       that.executeCallback("rejected", new TypeError(`Chaining cycle detected for promise #<Promise>`));
                    } else {
                       // 事件循环知识点需巩固: 比较巧妙  ③ ⑥
                        that.executeCallback("fulfilled", res);
                    }
                });
            }

            then(onResolved, onRejected) {
                if ((typeof (onResolved) !== "function") && this.state === "fulfilled" ||
                    (typeof (onRejected) !== "function") && this.state === "rejected") {
                    // 配合 done() 食用
                    return this;
                }
                // 创建一个新的 promise 实例。 作用一: 链式调用。 作用二: 传进
                // CallbackItem 中，使其能调用 Promise 的方法
                // new this.constructor() 是个什么意思？？？
                /**
                 * 详细示例:《深入理解ES6》\9th chapter--js中的类\es6中的类-class.js
                 * function Square(length) {
                 *     console.log("What is the Square this ?: ", this);
                 *     Rectangle.call(this, length, length);
                 * }

                 * Square.prototype = Object.create(Rectangle.prototype, {
                 *     constructor: {
                 *         // 这种写法等于: Square.prototype.constructor = Square
                 *         value: Square,
                 *         enumerable: true,
                 *         writable: true,
                 *         configurable: true
                 *     }
                 * });
                 * let square = new Square(3);
                 * console.log(square.getArea());              // 9
                 * console.log(square instanceof Square);      // true
                 * console.log(square instanceof Rectangle);   // true
                 */
                console.log("this: ", this);    // Promise
                console.log("this.constructor: ", this.constructor);
                // 这里既是: const promise = new Promise();
                const promise = new this.constructor();
                // 第一次进入 then, 状态时 fulfilled 或 rejected
                if (this.state !== "pending") {
                    const callback = this.state === "fulfilled" ? onResolved : onRejected;
                    // 绑定 this 到 promise   ①
                    this.executeAsyncCallback.call(promise, callback, this.data);
                }
                // 从第二次开始以后，进入 then 状态时 pending
                else {
                    // 这里的 this 也是指向 "上一个" promise  ②
                    this.callbackArr.push(new CallbackItem(promise, onResolved, onRejected));
                }
                return promise;
            }



            // 获取 thenable 对象
            getThen(value) {
                // 避免 then 内 get 方法多次的调用。 对应 /** # 特殊测试.html 调用 */
                const then = value.then;
                if (Object.prototype.toString.call(value) === "[object Object]" &&
                    typeof(then) === "function") {
                    const that = this;
                    return function() {
                        // 调用 then 对象，arguments 为 onSuccess 和 onError 函数
                        then.apply(value, arguments);
                    }
                } else {
                    return false;
                }
            }

            catch(onRejected) {
                // 加上 return, 相当于返回 promise
                return this.then(null, onRejected);
            }

            // 非 ES6 标准
            done(onResolved, onRejected) {
                this.then(onRsolved, onRejected).catch((reason) => {
                    setTimeout(() => {
                        throw reason;
                    }, 0)
                })
            }

        }


        class CallbackItem {
            constructor(promise, onResolved, onRejected) {
                // 相应地， 这里存储的 promise 是来自下一个 then 的
                this.promise = promise;
                this.onResolved = typeof (onResolved) === "function" ? onResolved : (value) => {
                    return value;
                };
                this.onRejected = typeof (onRejected) === "function" ? onRejected : (err) => {
                    throw err;
                }
            }
            fulfilled(value) {
                //  ⑤
                this.promise.executeAsyncCallback(this.onResolved, value);
            }
            rejected(value) {
                this.promise.executeAsyncCallback(this.onRejected, value);
            }
        }

        Promise.resolve = function(value) {
            // 在 Promise.race 中用到，使 Promise 对象: Promise.resolve(1) 和 普通值: 3 之间公平竞争。
            // 原理: 避免下一步进入 setTimeout 回调
            if (value instanceof this) return value;
            return this.prototype.executeCallback.call(new Promise(), "fulfilled", value);
        };

        Promise.reject = function(value) {
            if (value instanceof this) return value;
            return this.prototype.executeCallback.call(new Promise(), "rejected", value);
        };




        let p = new Promise (
            // function 函数传入到 Promise 对象中就是 resolver 函数
            function (resolve, reject) {
                setTimeout(function() {
                    resolve(0);
                    reject(1);
                }, 500)
            },
        );
        p.then(
            (data) => {
                console.log(data);
                return 2;
            }
        ).then(
            (data) => {
                console.log(data);
                return 3;
            }
        );


        /** # 1. 基础测试  */
        function pms1() {
            return new Promise(function (resolve, reject) {
                console.log('执行任务1');
                resolve('执行任务1成功')
            })
        }

        pms1().then(function (data) {
                console.log(`第一个回调：${data}`)
            });
        // 执行任务1
        // 第一个回调：执行任务1成功



        /** # 4 test4.html : resolve(Promise.resolve(1)) + 连续 then 调用 */
        function pms4() {
            return new Promise(function(resolve, reject) {
                console.log("执行任务4");
                resolve("执行任务4成功");
            })
        }
        function pms44() {
            return new Promise(function(resolve, reject) {
                console.log("执行任务44");
                resolve(Promise.resolve("执行任务44成功"));
            })
        }
        pms4().then(function(data) {
            console.log(data);
            return pms44();
        }).then(function(data) {
            console.log(data);
        });



        /** # 特殊测试:  */
        let numberOfTimesThenWasRetrieved = 0;
        /*
         * 注: ES5 - 提供了 Object.create 方法，可以用来克隆对象。
         * Object.create("要克隆的对象", "新对象定义额外属性的对象(可选,一般不写)")
         * 《js高程》P170 --> ECMAScript 5 通过新增 Object.create() 方法规范化了原型式继承。这个方法接收两个参数：一
         * 个用作新对象原型的对象和（可选的）一个为新对象定义额外属性的对象。在传入一个参数的情况下， Object.create() 与 object() 方法的行为相同。
         */
        // 利用 Object.create() 创建一个空对象，此对象上定义了一个 then 方法
        let obj = Object.create(null, {
            then: {
                get: function () {
                    ++numberOfTimesThenWasRetrieved;
                    return function(resolve) {
                        resolve();
                    }
                }
            }
        });
        Promise.resolve(obj)
            .then(() => {
                console.log(numberOfTimesThenWasRetrieved) // 1
            })

    </script>
</body>
</html>
